src/mod.test.ts

```diff
@@ -0,0 +1,487 @@
+    );
+
+    try {
+      await plugin.call('run_test', '');
+      assert.deepEqual(intercept, {
+        debug: 'this is a debug log',
+        error: 'this is an erorr log',
+        info: 'this is an info log',
+        warn: 'this is a warning log',
+      });
+    } finally {
+      await plugin.close();
+    }
+  });
+
+  test('host functions may read info from context and return values', async () => {
+    let executed: any;
+    const functions = {
+      env: {
+        hello_world(context: CallContext, off: bigint) {
+          executed = context.read(off)?.string();
+          return context.store('wow okay then');
+        },
+      },
+    };
+    const plugin = await createPlugin(
+      { wasm: [{ url: 'http://localhost:8124/wasm/code-functions.wasm' }] },
+      { useWasi: true, functions },
+    );
+
+    try {
+      const output = await plugin.call('count_vowels', 'hello world');
+      assert.equal(output?.string(), 'wow okay then');
+      assert.equal(executed, '{"count": 3}');
+    } finally {
+      await plugin.close();
+    }
+  });
+
+  test('resetting the plugin unsets all existing pages', async () => {
+    const offsets: bigint[] = [0n, 0n];
+    let callContext: CallContext | null = null;
+
+    const functions = {
+      env: {
+        hello_world(context: CallContext, off: bigint) {
+          callContext = context;
+
+          offsets[0] = off;
+          offsets[1] = context.store('wow okay then');
+          return offsets[1];
+        },
+      },
+    };
+    const plugin = await createPlugin(
+      { wasm: [{ url: 'http://localhost:8124/wasm/code-functions.wasm' }] },
+      { useWasi: true, functions },
+    );
+
+    try {
+      const output = await plugin.call('count_vowels', 'hello world');
+      assert.equal(output?.string(), 'wow okay then');
+
+      await plugin.reset();
+
+      assert(callContext !== null);
+      assert.notEqual(offsets[0], 0n);
+      assert.notEqual(offsets[1], 0n);
+      assert.equal((callContext as CallContext).read(offsets[0]), null);
+      assert.equal((callContext as CallContext).read(offsets[1]), null);
+    } finally {
+      await plugin.close();
+    }
+  });
+
+  test('host functions reject original promise when throwing', async () => {
+    const expected = String(Math.random());
+    const functions = {
+      env: {
+        hello_world(_context: CallContext, _off: bigint) {
+          throw new Error(expected);
+        },
+      },
+    };
+    const plugin = await createPlugin(
+      { wasm: [{ url: 'http://localhost:8124/wasm/code-functions.wasm' }] },
+      { useWasi: true, functions },
+    );
+
+    try {
+      const [err, data] = await plugin.call('count_vowels', 'hello world').then(
+        (data) => [null, data],
+        (err) => [err, null],
+      );
+
+      assert(data === null);
+      assert.equal(err?.message, expected);
+    } finally {
+      await plugin.close();
+    }
+  });
+
+  test('plugin can get/set variables', async () => {
```

Comment:
Bringing our chat from earlier into a comment here.  Not required to merge, but it will be good to include tests that ensure that plugin variables get persisted across multiple `plugin.call`s